### DISCUSSION TITLE
ARM support?

### DISCUSSION THREAD

**Discussion by Kjeldgaard** [lukevella/rallly#731] [General]
Hi,
Are there any plans to start building for ARM again? I remember around April you encountered problems building for ARM, so I'm still stuck on version 2.4.0(?).

**Comment by lukevella** [comment]
Hi, I definitely would like to but unfortunately running the builds on Github x86 runners causes too many issues and I can't justify paying out of pocket for dedicated ARM runners. I tried to get sponsorship for ARM runners but didn't have any luck with that. I'm hoping some sort of support will come through that will allow me to publish ARM builds again but in the mean time it won't be possible.

**Reply by JavierLuna** [reply]
Out of curiosity, what issues do the x86 runners cause when building arm images? I was under the impression that docker buildx handled everything.

Anything I could help with? Just tried self-hosting rallly and stumbled across the lack of support 😓 would like to help if that's possible.

**Reply by lukevella** [reply]
The issue with using x86 runners is that the build is very slow (over an hour) and tends to fail due to timing out quite often. I tried to resolve the timeout issue which I recall was coming from yarn but it was still failing regularly so I gave up on it. It might be worth trying again to be fair since a lot has changed in the codebase.

**Reply by colfes** [reply]
Hi there, I just tried it again. You can find a pipeline at fius/rallly, the container for v.3.1.1 ist also pushed to fius/rallly at hub.docker.com.
It took almost 2 hours using the GitHub runner, on a "normal" desktop machine, it's not really feasible. I've quickly testet it on a Raspberry Pi 3b+ and it's working fine.

**Comment by colfes** [comment]
Added an ARM-build for v3.2.1 (https://hub.docker.com/r/fius/rallly)

**Comment by tisba** [comment]
FYI: ARM based runners are in beta https://github.blog/changelog/2023-10-30-accelerate-your-ci-cd-with-arm-based-hosted-runners-in-github-actions/

BUT I think what could also work in the meantime is to use `macos-14`, which are exclusively run on M1 Macs (see https://github.blog/changelog/2024-01-30-github-actions-macos-14-sonoma-is-now-available/). Unless you totally oppose @lukevella, I can give it a try and send a PR if it works reasonably well.

**Reply by lukevella** [reply]
Yes, would totally welcome a PR for that 👍

**Reply by tisba** [reply]
arg, damn, totally forgot about the need and lack of support for nested virtualisation 🙄  https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners/about-larger-runners#limitations-for-macos-larger-runners:

> Due to a limitation of Apple's Virtualization Framework, which our hypervisor uses, nested-virtualization is not supported by arm64 runners.

So I guess in the current state, we have to wait.

Another option, for which I know too little about the current/modern Node.js build systems, is to make the Docker builds cachable. I noticed that the resulting image is **massive** in size (~1.4GB). The workflow currently makes no use of caching at all. Even if the QEMU-based builds won't be super fast, it might be possible to significantly reduce the build time. Not sure if it's worth the work for the project though.

**Reply by lukevella** [reply]
I've given up on the QEMU-based build as they are just too flaky. Even if it works for some time, eventually the cache will get invalidated and I can't justify spending time working around the resulting issues. I've signed up for access to arm runners so hopefully I'll hear back and that could be a way forward.

**Reply by tisba** [reply]
Hey @lukevella, I'm curious: Did you get into the ARM runner early access / private beta? In case you got access, I'd happy to lend you a hand getting this up and running if you like.

**Reply by lukevella** [reply]
@tisba Still on the waitlist unfortunately but I can let you know once I'm in. Would appreciate help with this one 👍

**Reply by tisba** [reply]
Just a quick update here: Linux ARM runners are in public beta for Github Enterprise customers. Announced per https://github.blog/changelog/2024-06-03-actions-arm-based-linux-and-windows-runners-are-now-in-public-beta/:

> We expect to begin offering Arm runners for open source and personal accounts by the end of the year.

Would love to get this sooner of course, but that's at least something 🤞

**Reply by tisba** [reply]
Here we go, arm64 Linux runners for free in public repos:

https://github.blog/changelog/2025-01-16-linux-arm64-hosted-runners-now-available-for-free-in-public-repositories-public-preview/

I'll try to find some time in the next days to get this setup for rallly

**Reply by edoren** [reply]
Hello! I went ahead and created this PR to add support to creating ARM64 images, some reviews would be appreciated 😄 
https://github.com/lukevella/rallly/pull/1532
